mathics/autoload/rules/Limit.m

```diff
@@ -0,0 +1,22 @@
+(* Additions to mathics.builtin.numbers.trig that are either wrong
+   or not covered by SymPy.
+
+   These were culled from symja_android_library/rules/LimitRules.m.
+ *)
+
+Unprotect[Limit];
+(*****
+ Fixme  Uncommenting any of the below causes:
+
+  G[x_Real]=x^2; a={G[x]}; {x=1.; a, x=.; a} == {{1.}, {G[x]}}
```

Comment:
I guess the reason is because of the context of the patterns. What about using 
```
Limit[Tan[System`limits`x_], System`limits`x_Symbol->Pi/2]=Indeterminate;
```
?

Reply:
I assume you meant `System`Limit`x_` ?  That seems to work. 

Does WMA require this kind of thing too? I would find it hard to believe because I see many packages out there that use short names without context qualification. 

And exactly why does changing one cause something that seems totally unrelated break?

Reply:
There are two different things. On the one hand, in WMA packages, the definitions are embraced on a ```Begin["context`"]/End[] ```
This prevents `x` and other symbols associated with patterns from being stored in the ```Global` ``` context.

The other thing is maybe related to this:

https://github.com/Mathics3/mathics-core/blob/49ed7c9f8ba08722836bc415dcd584ac96c91dba/mathics/core/definitions.py#L59

We set the context in autoload packages to ```System` ``` by default. Then, ```x``` is interpreted as ```System`x ``` instead of  ``` Global`x  ```, which hides this other definition.

Reply:
Again, my question is - is this how WMA works? 

In particular - 

Does WMA have a mechanism whereby the user can load definitions that change the behavior like add rules like this to Built-in functions? 

If it does have a mechanism, does it also requery specifying  the "System" (and other nested) contexts? 

If it doesn't then we have a bug in Mathics3 somewhere.

Reply:
In Mathematica, by default, the Context used when Autoload packages are loaded is ``` Global` ```. Also, by looking at what happens if I define a variable in an autoload module and I clear it afterward, I would say that these definitions are not moved into the "builtin" definitions, as we does in Mathics.

For example, if I add the following file to the autoload folder:
``` (*  ~/Autoload/TestAutoload/init.m  *)

Print[$Context]

x="Hola"
```


each time I start a WMA REPL, ```  Global`   ```  is printed before the prompt 
```
 In[1]:=
```
Now, if I evaluate `x`,  I get
```
 In[1]:=x
Out[1] Hola
```
and 
```
 In[2]:=? x

Out[2]= Global`x
        x = "Hola"

```
Then,
```
 In[3]:= Clear[x]; x

Out[3]= x
```

Reply:
> In Mathematica, by default, the Context used when Autoload packages are loaded is `` Global` ``. Also, by looking at what happens if I define a variable in an autoload module and I clear it afterward, I would say that these definitions are not moved into the "builtin" definitions, as we does in Mathics.
> 
> For example, if I add the following file to the autoload folder:
> 
> ```
>  (*  ~/Autoload/TestAutoload/init.m  *)
> 
> Print[$Context]
> 
> x="Hola"
> ```
> 
> each time I start a WMA REPL, `` Global` `` is printed before the prompt
> 
> ```
>  In[1]:=
> ```
> 
> Now, if I evaluate `x`, I get
> 
> ```
>  In[1]:=x
> Out[1] Hola
> ```
> 
> and
> 
> ```
>  In[2]:=? x
> 
> Out[2]= Global`x
>         x = "Hola"
> ```
> 
> Then,
> 
> ```
>  In[3]:= Clear[x]; x
> 
> Out[3]= x
> ```

Thanks for the explanation. So as I understand it, the "limits" in ``System`limits`x_`` was correct in that we just needed a new name that is not likely to be used elsewhere. "Limit" matches the builtin name, but all that matters is that it is not likely to be used elsewhere.

I guess putting things in System is a better or more protected than Global, and this awkwardness is limited just to predefined builtin functions. So I guess the way things are is good. For all we know Mathematics _does_ do a copy for these autoloaded rules that you do not see.